ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/genesis/GenesisHandler.java

```diff
@@ -0,0 +1,117 @@
+import static tech.pegasys.artemis.datastructures.util.BeaconStateUtil.is_valid_genesis_state;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.artemis.datastructures.operations.DepositWithIndex;
+import tech.pegasys.artemis.datastructures.state.BeaconStateWithCache;
+import tech.pegasys.artemis.datastructures.util.DepositUtil;
+import tech.pegasys.artemis.datastructures.util.GenesisGenerator;
+import tech.pegasys.artemis.pow.api.DepositEventChannel;
+import tech.pegasys.artemis.pow.api.MinGenesisTimeBlockEventChannel;
+import tech.pegasys.artemis.pow.event.DepositsFromBlockEvent;
+import tech.pegasys.artemis.pow.event.MinGenesisTimeBlockEvent;
+import tech.pegasys.artemis.statetransition.events.GenesisEvent;
+import tech.pegasys.artemis.storage.ChainStorageClient;
+
+public class GenesisHandler implements DepositEventChannel, MinGenesisTimeBlockEventChannel {
+
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final ChainStorageClient chainStorageClient;
+  private final GenesisGenerator genesisGenerator = new GenesisGenerator();
+
+  private Queue<DepositsFromBlockEvent> bufferedDepositsFromBlockEvents = new LinkedList<>();
+  private volatile boolean genesisAlreadyTriggered = false;
+
+  public GenesisHandler(final ChainStorageClient chainStorageClient) {
+    this.chainStorageClient = chainStorageClient;
+  }
+
+  @Override
+  public synchronized void onDepositsFromBlock(final DepositsFromBlockEvent event) {
+    if (!chainStorageClient.isPreGenesis() || genesisAlreadyTriggered) {
+      return;
+    }
+
+    if (isThereEnoughNumberOfValidators(genesisGenerator.getCandidateState())) {
+      bufferedDepositsFromBlockEvents.add(event);
+    } else {
+      addDepositsToState(event);
+      if (is_valid_genesis_state(genesisGenerator.getCandidateState())) {
+        triggerGenesis();
+      }
+    }
+  }
+
+  private void addDepositsToState(DepositsFromBlockEvent event) {
+    final Bytes32 eth1BlockHash = event.getBlockHash();
+    final UnsignedLong eth1Timestamp = event.getBlockTimestamp();
+    final List<DepositWithIndex> deposits =
+        event.getDeposits().stream()
+            .map(DepositUtil::convertDepositEventToOperationDeposit)
+            .collect(Collectors.toList());
+    genesisGenerator.addDepositsFromBlock(eth1BlockHash, eth1Timestamp, deposits);
+  }
+
+  @Override
+  public synchronized void onMinGenesisTimeBlock(MinGenesisTimeBlockEvent event) {
+    if (genesisAlreadyTriggered) {
+      return;
+    }
+
+    if (isThereEnoughNumberOfValidators(genesisGenerator.getCandidateState())) {
+      processEventsUpToMinGenesisTimeBlock(event);
+      triggerGenesis();
+    }
```

Comment:
This turns out to be incorrect and I think we just broke our whole algorithm (again, sigh).  Suppose we get the events:

 - 16,000 validators register in block 10 - we now have enough validators registered
 - Min genesis block received specifying block 50 is min genesis.  We will trigger genesis at this point
 - 100 validators register in block 11.  We should have included them in the genesis (they're prior to the min genesis block)

The only way I can see to solve this is to do things in the right order back in the PowService.  So I think it should:
 1. Find the minimum genesis block.
 2. Request and send deposits between 0 & min genesis block
 3. Send event notifying that minimum genesis time has now been reached
 4. Request and send deposits after the min genesis block

The deposit and min genesis notification would need to be on the same channel to ensure they maintain the correct order.

If min genesis time hasn't yet been reached by the ETH1 chain (considering follow distance) we just wind up tracking head and sending the min genesis time reached event when it does get there.